src/lib.rs

```diff
@@ -534,6 +535,36 @@ impl<T: Abomonation> Abomonation for Box<T> {
     }
 }
 
+impl<T: Abomonation + 'static + Sized + Unpin> Abomonation for Arc<T> {
+    #[inline]
+    unsafe fn entomb<W: std::io::Write>(&self, write: &mut W) -> std::io::Result<()> {
+        /* Since the value T has not yet been seen by abomonate (it is behind a pointer)
+         * we need to fully encode it.
+         */
+        encode(self.as_ref(), write)
+    }
+    #[inline]
+    unsafe fn exhume<'a, 'b>(&'a mut self, bytes: &'b mut [u8]) -> Option<&'b mut [u8]> {
+        use std::ptr;
+        /* The idea here is to construct a new Arc<T> from the entombed bytes.
+         * The state of this ArcVal upon entry of this function contains only an invalid
+         * pointer to an ArcInner that we need to dispose of without trying to run
+         * its destructor (which would panic).
+         */
+        let (value, bytes) = decode::<T>(bytes)?;
+        // value is just a reference to the first part of old bytes, so move it into a new Arc
+        let arc = Arc::new(ptr::read(value));
```

Comment:
Here we lift the bytes from the underlying slice and move them into the ArcInner, i.e. we move the value of type `T` to a new memory location. Since it cannot reference `bytes` by the above argument, the result should be a correctly self-contained `Arc<T>`.

Reply:
I think this might be the problematic moment. Something like `String` can still reference `bytes` even after a `ptr::read`. The read will only grab the 24 bytes of the `String` type, but it may contain a pointer in to `bytes` (and very likely does, as this is how `String` implements `Abomonation`).

Reply:
Wow, that’s evil! So the only correct implementation would be to require `T: Clone` and clone the referenced value returned by `decode`.